mcs/class/Mono.Messaging/Mono.Messaging/MessagingProviderLocator.cs

```diff
@@ -43,20 +43,25 @@ namespace Mono.Messaging
 	public class MessagingProviderLocator 
 	{
 		public static readonly TimeSpan InfiniteTimeout = TimeSpan.MaxValue;
-		private static readonly MessagingProviderLocator instance = new MessagingProviderLocator();
 		private readonly IMessagingProvider provider;
 		private const string MESSAGING_PROVIDER_KEY = "MONO_MESSAGING_PROVIDER";
 		private const string RABBIT_MQ_CLASS_NAME = "Mono.Messaging.RabbitMQ.RabbitMQMessagingProvider";
 		private const string RABBIT_MQ_FULL_CLASS_NAME = RABBIT_MQ_CLASS_NAME + ",Mono.Messaging.RabbitMQ";
 		private const string RABBIT_MQ_ALIAS = "rabbitmq";
-		
-		private MessagingProviderLocator () {
-			string providerName = GetProviderClassName ();
+
+		internal delegate string MessagingProviderStrategy ();
+		internal static MessagingProviderStrategy strategy = GetProviderClassName;
```

Comment:
I don't really get why you need this delegate solution? Can you share some more details?